### PR TITLE
SLVS-982 Normalize symbolic values regarding (in)equality relationships

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

File: src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/Binary/XorSymbolicValue.cs [hunk 1]
```diff
@@ -18,7 +18,6 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
  */
 
-using System;
```
Comment (m-g-sonar) [inline src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/Binary/XorSymbolicValue.cs]: this change has nothing to do with your commit.

File: src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/Binary/AndSymbolicValue.cs [hunk 2]
```diff
@@ -18,7 +18,6 @@
  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
  */
 
-using System;
```
Comment (m-g-sonar) [inline src/SonarLint/Helpers/SymbolicExecution/SymbolicValues/Binary/AndSymbolicValue.cs]: this change has nothing to do with your work on this ticket.

File: src/Tests/SonarLint.UnitTest/TestCases/ConditionEvaluatesToConstant.cs [hunk 3]
```diff
@@ -438,10 +438,14 @@ public void EqRelations(bool a, bool b, object o)
             if (a == b)
             {
                 if (b == a) { } // Noncompliant {{Change this condition so that it does not always evaluate to "true".}}
+                if (b == !a) { }   // Noncompliant {{Change this condition so that it does not always evaluate to "false".}}
+                if (!b == !!a) { } // Noncompliant {{Change this condition so that it does not always evaluate to "false".}}
```
Comment (m-g-sonar) [inline src/Tests/SonarLint.UnitTest/TestCases/ConditionEvaluatesToConstant.cs]: You should add `!(a == b)` also, which should be always `false`.

File: src/SonarLint/Helpers/SymbolicExecution/Relationships/EqualsRelationship.cs [hunk 4]
```diff
@@ -18,43 +18,38 @@
...
-            return relationships
-                .OfType<NotEqualsRelationship>()
-                .Any(rel => OperandsMatch(rel));
-        }
-
-        public override string ToString()
+        protected bool Equals(EqualsRelationship other)
         {
-            return $"{LeftOperand} == {RightOperand}";
+            return other != null && OperandsMatch(other);
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/Relationships/EqualsRelationship.cs]: MAJOR Consider changing the type of parameter 'other' in 'EqualsRelationship.Equals(EqualsRelationship)' from 'EqualsRelationship' to its base type 'BinaryRelationship'. This method appears to only require base class members in its implementation. Suppress this violation if there is a compelling reason to require the more derived type in the method signature. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AConsiderPassingBaseTypesAsParameters)

File: src/SonarLint/Helpers/SymbolicExecution/Relationships/NotEqualsRelationship.cs [hunk 5]
```diff
@@ -18,45 +18,26 @@
...
         {
         }
 
-        internal override bool IsContradicting(IEnumerable<BinaryRelationship> relationships)
+        protected bool Equals(NotEqualsRelationship other)
         {
-            return relationships
-                .OfType<EqualsRelationship>()
-                .Any(rel => OperandsMatch(rel));
+            return other != null && OperandsMatch(other);
```
Comment (SonarTech) [inline src/SonarLint/Helpers/SymbolicExecution/Relationships/NotEqualsRelationship.cs]: MAJOR Consider changing the type of parameter 'other' in 'NotEqualsRelationship.Equals(NotEqualsRelationship)' from 'NotEqualsRelationship' to its base type 'BinaryRelationship'. This method appears to only require base class members in its implementation. Suppress this violation if there is a compelling reason to require the more derived type in the method signature. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AConsiderPassingBaseTypesAsParameters)

**Comment by SonarTech** [comment]
SonarQube analysis reported 8 issues
* MAJOR 8 major

Watch the comments in this conversation to review them.

#### 6 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: 'ReferenceEqualsSymbolicValue' has an object hierarchy 6 levels deep within the defining module. If possible, eliminate base classes within the hierarchy: 

SonarLint.Helpers.FlowAnalysis.Common.ReferenceEqualsSymbolicValue
  SonarLint.Helpers.FlowAnalysis.Common.EqualsSymbolicValue
   SonarLint.Helpers.FlowAnalysis.Common.EqualityLikeSymbolicValue
    SonarLint.Helpers.FlowAnalysis.Common.RelationalSymbolicValue
     SonarLint.Helpers.FlowAnalysis.Common.BinarySymbolicValue
      SonarLint.Helpers.FlowAnalysis.Common.SymbolicValue [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidExcessiveInheritance)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: 'ReferenceNotEqualsSymbolicValue' has an object hierarchy 6 levels deep within the defining module. If possible, eliminate base classes within the hierarchy: 

SonarLint.Helpers.FlowAnalysis.Common.ReferenceNotEqualsSymbolicValue
  SonarLint.Helpers.FlowAnalysis.Common.NotEqualsSymbolicValue
   SonarLint.Helpers.FlowAnalysis.Common.EqualityLikeSymbolicValue
    SonarLint.Helpers.FlowAnalysis.Common.RelationalSymbolicValue
     SonarLint.Helpers.FlowAnalysis.Common.BinarySymbolicValue
      SonarLint.Helpers.FlowAnalysis.Common.SymbolicValue [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidExcessiveInheritance)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'ValueEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: 'ValueEqualsSymbolicValue' has an object hierarchy 6 levels deep within the defining module. If possible, eliminate base classes within the hierarchy: 

SonarLint.Helpers.FlowAnalysis.Common.ValueEqualsSymbolicValue
  SonarLint.Helpers.FlowAnalysis.Common.EqualsSymbolicValue
   SonarLint.Helpers.FlowAnalysis.Common.EqualityLikeSymbolicValue
    SonarLint.Helpers.FlowAnalysis.Common.RelationalSymbolicValue
     SonarLint.Helpers.FlowAnalysis.Common.BinarySymbolicValue
      SonarLint.Helpers.FlowAnalysis.Common.SymbolicValue [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidExcessiveInheritance)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: Since 'ValueNotEqualsRelationship' redefines Equals, it should also redefine GetHashCode. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AOverrideGetHashCodeOnOverridingEquals)
1. MAJOR sonaranalyzer-csharp-vbnet:sonaranalyzer-csharp-vbnet:D1AE804F-AE78-4883-B3D6-9E3C4026DEF6: 'ValueNotEqualsSymbolicValue' has an object hierarchy 6 levels deep within the defining module. If possible, eliminate base classes within the hierarchy: 

SonarLint.Helpers.FlowAnalysis.Common.ValueNotEqualsSymbolicValue
  SonarLint.Helpers.FlowAnalysis.Common.NotEqualsSymbolicValue
   SonarLint.Helpers.FlowAnalysis.Common.EqualityLikeSymbolicValue
    SonarLint.Helpers.FlowAnalysis.Common.RelationalSymbolicValue
     SonarLint.Helpers.FlowAnalysis.Common.BinarySymbolicValue
      SonarLint.Helpers.FlowAnalysis.Common.SymbolicValue [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=fxcop%3AAvoidExcessiveInheritance)

**Pull request merged**